### PR TITLE
[SLP][REVEC] Make reorderTopToBottom support ShuffleVectorInst.

### PR DESCRIPTION
**Pull request by HanKuanChen**
We don't want reorderTopToBottom to reorder ShuffleVectorInst (because ShuffleVectorInst currently supports only a limited set of patterns). Either we make ShuffleVectorInst support more patterns, or we let ReorderIndices reorder the result of the vectorization of ShuffleVectorInst. We choose the latter solution.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -6005,10 +6005,12 @@ void BoUpSLP::reorderTopToBottom() {
       if ((TE->State == TreeEntry::Vectorize ||
            TE->State == TreeEntry::StridedVectorize) &&
           isa<ExtractElementInst, ExtractValueInst, LoadInst, StoreInst,
-              InsertElementInst>(TE->getMainOp())) {
+              InsertElementInst, ShuffleVectorInst>(TE->getMainOp())) {
         assert(!TE->isAltShuffle() &&
                "Alternate instructions are only supported by BinaryOperator "
                "and CastInst.");
+        assert(!isa<ShuffleVectorInst>(TE->getMainOp()) ||
+               SLPReVec && "Only supported by REVEC.");
```
Comment (alexey-bataev) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:6013]: It should be not an assertion, but a logic in the if condition

Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:6013]: I don't get it. Only REVEC can support `ShuffleVectorInst`, why do not use assert here?

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:6013]: Without revec the compiler should not crash on shuffles, it just shall ignore them

Comment (HanKuanChen) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:6013]: How does the TreeEntry has `ShuffleVectorInst` as Scalars without REVEC? `ShuffleVectorInst` returns vector.

Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:6013]: It may help for constant shuffles in some cases, I assume

**Review by alexey-bataev — APPROVED** [review]

**Pull request merged**